File: src/main/java/idlezoo/game/services/ResourcesService.java
package idlezoo.game.services;

import static java.util.Collections.unmodifiableList;
import static java.util.Collections.unmodifiableMap;

import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.springframework.beans.factory.InitializingBean;
import org.springframework.stereotype.Service;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.type.CollectionType;

import idlezoo.game.domain.Building;
import idlezoo.game.domain.Zoo;
import idlezoo.game.domain.Perks.Perk;
import one.util.streamex.StreamEx;

@Service
public class ResourcesService implements InitializingBean {

    private final ObjectMapper mapper;

    private double startingMoney = 50D;
    private List<Building> animalsList;
    private List<Perk> perkList;
    private Map<String, Integer> animalIndexes = new HashMap<>();
    private Map<String, Building> animalTypes = new HashMap<>();
    private Map<String, Building> nextAnimals = new HashMap<>();
    private Map<String, Integer> perkIndexes = new HashMap<>();
    private Map<String, Perk> perkNames = new HashMap<>();
    private Building startingAnimal;

    public ResourcesService(ObjectMapper mapper) {
        this.mapper = mapper;
    }

    @Override
    public void afterPropertiesSet() throws IOException {
        initAnimals();
        initPerks();
    }

    private void initPerks() throws IOException {
        try (InputStream perks = ResourcesService.class.getResourceAsStream(
                "/mechanics/perks.json")) {
            CollectionType type = mapper.getTypeFactory().constructCollectionType(List.class, Perk.class);
            perkList = unmodifiableList(mapper.readValue(perks, type));
            for (int i = 0; i < perkList.size(); i++) {
                Perk perk = perkList.get(i);
                perkIndexes.put(perk.getName(), i);
                perkNames.put(perk.getName(), perk);
            }
            perkIndexes = unmodifiableMap(perkIndexes);
            perkNames = unmodifiableMap(perkNames);
        }
    }

    private void initAnimals() throws IOException {
        try (InputStream creatures = ResourcesService.class.getResourceAsStream(
                "/mechanics/animals.json")) {
            CollectionType type = mapper.getTypeFactory().constructCollectionType(List.class,
                    Building.class);
            animalsList = unmodifiableList(mapper.readValue(creatures, type));
            startingAnimal = animalsList.get(0);

            Building prev = null;
            for (int i = 0; i < animalsList.size(); i++) {
                Building animal = animalsList.get(i);
                animalIndexes.put(animal.getName(), i);
                animalTypes.put(animal.getName(), animal);
                if (prev != null) {
                    nextAnimals.put(prev.getName(), animal);
                }
                prev = animal;
            }
        }
        animalTypes = unmodifiableMap(animalTypes);
        nextAnimals = unmodifiableMap(nextAnimals);
        animalIndexes = unmodifiableMap(animalIndexes);
    }

    double startingMoney() {
        return startingMoney;
    }

    Building startingAnimal() {
        return startingAnimal;
    }

    String firstName() {
        return startingAnimal.getName();
    }

    public String secondName() {
        return nextType(firstName()).getName();
    }

    public Building nextType(String buildingName) {
        return nextAnimals.get(buildingName);
    }

    public Building type(String typeName) {
        return animalTypes.get(typeName);
    }

    public List<Building> getAnimalsList() {
        return animalsList;
    }

    public List<Perk> getPerkList() {
        return perkList;
    }

    public Perk perkByIndex(int index) {
        return perkList.get(index);
    }

    public Perk perk(String name) {
        return perkNames.get(name);
    }

    public Integer perkIndex(String perkName) {
        return perkIndexes.get(perkName);
    }

    public Building animalByIndex(int index) {
        return animalsList.get(index);
    }

    public Integer animalIndex(String animalName) {
        return animalIndexes.get(animalName);
    }

    public List<Perk> availablePerks(Zoo zoo) {
        List<Perk> result = new ArrayList<>(perkList);
        result.removeAll(zoo.getPerks());
        return StreamEx.of(result)
                .filter(perk -> perk.isAvailable(zoo))
                .toList();
    }

}

File: src/main/kotlin/idlezoo/game/services/ResourcesService.kt
package idlezoo.game.services

import com.fasterxml.jackson.databind.ObjectMapper
import idlezoo.game.domain.Building
import idlezoo.game.domain.Perks.Perk
import idlezoo.game.domain.Zoo
import one.util.streamex.StreamEx
import org.springframework.beans.factory.InitializingBean
import org.springframework.stereotype.Service
import java.util.*
import java.util.Collections.unmodifiableList

@Service
class ResourcesService(private val mapper: ObjectMapper) : InitializingBean {

    lateinit var animalsList: List<Building>
    lateinit var perkList: List<Perk>
    private lateinit var animalIndexes: Map<String, Int>
    private lateinit var animalTypes: Map<String, Building>
    private lateinit var nextAnimals: Map<String, Building>
    private lateinit var perkIndexes: Map<String, Int>
    private lateinit var perkNames: Map<String, Perk>
    private lateinit var startingAnimal: Building

    override fun afterPropertiesSet() {
        initAnimals()
        initPerks()
    }

    private fun initPerks() {
        val perkIndexesBuilder = mutableMapOf<String, Int>()
        val perkNamesBuilder = mutableMapOf<String, Perk>()
        ResourcesService::class.java.getResourceAsStream(
                "/mechanics/perks.json").use { perks ->
            val type = mapper.typeFactory.constructCollectionType(List::class.java, Perk::class.java)
            perkList = unmodifiableList(mapper.readValue(perks, type))
            for (i in perkList.indices) {
                val perk = perkList[i]
                perkIndexesBuilder[perk.name] = i
                perkNamesBuilder[perk.name] = perk
            }
        }
        perkIndexes = perkIndexesBuilder
        perkNames = perkNamesBuilder
    }

    private fun initAnimals() {
        val animalTypesBuilder = mutableMapOf<String, Building>()
        val nextAnimalsBuilder = mutableMapOf<String, Building>()
        val animalIndexesBuilder = mutableMapOf<String, Int>()
        ResourcesService::class.java.getResourceAsStream(
                "/mechanics/animals.json").use { creatures ->
            val type = mapper.typeFactory.constructCollectionType(List::class.java,
                    Building::class.java)
            animalsList = unmodifiableList(mapper.readValue(creatures, type))
            startingAnimal = animalsList[0]

            var prev: Building? = null
            for (i in animalsList.indices) {
                val animal = animalsList[i]
                animalIndexesBuilder[animal.name] = i
                animalTypesBuilder[animal.name] = animal
                if (prev != null) {
                    nextAnimalsBuilder[prev.name] = animal
                }
                prev = animal
            }
        }
        animalTypes = animalTypesBuilder
        nextAnimals = nextAnimalsBuilder
        animalIndexes = animalIndexesBuilder
    }

    internal fun startingMoney(): Double {
        return STARTING_MONEY
    }

    internal fun startingAnimal(): Building? {
        return startingAnimal
    }

    internal fun firstName(): String {
        return startingAnimal.name
    }

    fun secondName(): String {
        return nextType(firstName())!!.name
    }

    fun nextType(buildingName: String): Building? {
        return nextAnimals[buildingName]
    }

    fun type(typeName: String): Building {
        return animalTypes[typeName]!!
    }

    fun perkByIndex(index: Int): Perk {
        return perkList[index]
    }

    fun perk(name: String): Perk {
        return perkNames[name]!!
    }

    fun perkIndex(perkName: String): Int? {
        return perkIndexes[perkName]
    }

    fun animalByIndex(index: Int): Building {
        return animalsList[index]
    }

    fun animalIndex(animalName: String): Int? {
        return animalIndexes[animalName]
    }

    fun availablePerks(zoo: Zoo): List<Perk> {
        val result = ArrayList(perkList)
        result.removeAll(zoo.perks)
        return StreamEx.of(result)
                .filter { perk -> perk.isAvailable(zoo) }
                .toList()
    }

    companion object {
        private val STARTING_MONEY = 50.0
    }
}
